Primerize the user profile page

oliverguenther opened the pull request:
Follow-up for https://community.openproject.org/work_packages/55602

https://community.openproject.org/work_packages/55605

oliverguenther committed:
Add groups to user profile

oliverguenther committed:
Primerize

Use inline erb

Primerize expandable list

oliverguenther committed:
Fix lookbook accessing behind proxy

oliverguenther committed:
Move spec to component

oliverguenther committed:
Changed condition 404 -> 200

The user can no be viewed, even if subsequent information on
groups/memberships are hidden

ulferts commented on app/components/users/profile/attributes_component.html.erb:
```diff
@@ -0,0 +1,37 @@
+      @user.visible_custom_field_values.each do |custom_value|
+        if custom_value.value.present?
+          details_container.with_row do
+            "#{custom_value.custom_field.name}: #{custom_value.formatted_value}"
```
This leads to the following problems:

* Long text custom fields are rendered escaped:
<img width="325" alt="image" src="https://github.com/opf/openproject/assets/617519/65fb8d66-5152-4991-a6a1-814953334e6d">

* Multi value list custom fields are rendered multiple times (existed before the PR as well):
<img width="103" alt="image" src="https://github.com/opf/openproject/assets/617519/8492ffff-6755-4258-b5a7-8d98f310b253">

It would be nice, if the fields were sorted alphabetically.

ulferts commented on app/components/users/profile/projects_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+        end
+      end
+    end
+  end
```
The component does not have a descriptive text /subheading and no counter pill as in the figma file:

<img width="239" alt="image" src="https://github.com/opf/openproject/assets/617519/20040842-b446-49a2-93f0-1e29108f33db">

oliverguenther replied:
We decided against the subheading, Figma is updated

ulferts commented on app/components/users/profile/groups_component.html.erb:
```diff
@@ -0,0 +1,32 @@
+        end
+      end
+    end
+  end
```
Same as for the projects sections, there is no subheading and no counter pill.

oliverguenther replied:
We decided against the subheading, Figma is updated

ulferts commented on app/components/users/profile/projects_component.rb:
```diff
@@ -0,0 +1,56 @@
+        # also gets applied to the Project.allowed_to parts concatenated by a UNION
+        # and an order inside a UNION is not allowed in postgres.
+        @memberships = @user.memberships
+                            .where.not(project_id: nil)
```
An entity membership also has a project_id set:
<img width="417" alt="image" src="https://github.com/opf/openproject/assets/617519/9472cbb8-800e-4793-b9f9-0b4bd4c6a737">

The `of_any_project` scope is best to use here.

ulferts commented on app/components/users/profile/activities_component.html.erb:
```diff
@@ -0,0 +1,14 @@
+  </p>
+  <%= render(Activities::DaysComponent.new(events: @events, display_user: false, header_tag: 'h4', activity_page: @activity_page)) %>
+  <%= helpers.other_formats_links do |f| %>
+    <%= f.link_to 'Atom', url: { controller: '/activities', action: 'index', id: nil, user_id: @user, key: User.current.rss_key } %>
```
I guess not i18n-ing 'Atom' is safe since it is the format same as 'PDF'.

oliverguenther committed:
Fix back link in activities

ulferts commented on app/components/op_primer/expandable_list_component.rb:
```diff
@@ -25,33 +27,29 @@
+    include OpPrimer::ComponentHelpers
 
-require "spec_helper"
+    CUTOFF_LIMIT = 10
```
The new cutoff limit is defined but I believe in all of the calls in the productive code, it is still overwritten with 5 which contradicts the specification in the work package.

ulferts commented on app/components/users/profile/attributes_component.rb:
```diff
@@ -0,0 +1,51 @@
+
+      def visible_user_information?
+        @user.pref.can_expose_mail? || @user.visible_custom_field_values.any? { _1.value.present? }
+      end
```
Pulling this into the `render` method would be cleaner. Not a big issue, though.

ulferts commented on app/components/users/profile/projects_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+
+              roles = membership.roles.sort.collect(&:to_s).join(', ')
+              date = helpers.format_date(membership.created_at)
+              concat render(Primer::Beta::Text.new(ml: 1, color: :subtle, font_size: :small)) { "(#{roles}, #{date})" }
```
The figma files does not specify the listing of the roles. I however would prefer having the roles. Don't know which information need the date is catering to, though.

oliverguenther replied:
The previous profile showed them, so we'll keep them as is

ulferts commented on app/components/users/profile/attributes_component.html.erb:
```diff
@@ -0,0 +1,37 @@
+                scheme: :secondary
+              )) { t(:label_manage) }
+            end
+          end
```
If the current user is not the admin but the displayed user, we could link to the account page.

Here, a 'manage' link is displayed whereas on the meetings sidebar, a gear-icon is used. I think we should settle on one of these.

ulferts reviewed (changes requested):
Code-wise this is definitely heading into the right direction. On the UX level, there are a number of things that could possible be improved but I don't know how far this page is to be taken now.

The issue that should definitely be fixed IMO is the topic about the long text custom fields.

oliverguenther committed:
Use other scope

oliverguenther committed:
Fix rendering of CFs

oliverguenther committed:
Use default cutoff

oliverguenther committed:
Change to icons and other feedback

oliverguenther committed:
Change display of attributes to match figma

oliverguenther committed:
Order projects by name

oliverguenther commented:
@ulferts I updated the pills and fixed rendering of the custom fields, including multi-value rendering. Could you take another look? Some additional changes discussed with Parimal:

- Manage links changed to gear
- Projects ordered by name ASC, removed date as it's not clear what it is
- Changed display of attributes to match project overview

ulferts reviewed (approved):
👍 all good
